Add optional cop2_notify callback to lightrec_ops

ZachCook opened the pull request:
The callback is given the full opcode and the data that was read or written to or from the cop2 register

The callback is expected to decode the opcode, for example to ignore LWC2/SWC2 and determine which of the MFC/CFC/MTC/CTC functions was called

This will be used to provide the opcode and return/data value to PGXP_GTE_[MFC/CFC/MTC/CTC] functions in beetle-psx that was lost when cop2 became internal to lightrec

Signed-off-by: Zachary Cook <zachcook1991@gmail.com>

pcercuei commented on lightrec.c:
```diff
@@ -428,22 +428,26 @@ u32 lightrec_mfc(struct lightrec_state *state, union code op)
+	}
 
-	val = state->regs.cp2c[op.r.rd];
+	if (*state->ops.cop2_pgxp)
```
No need to deference - `if (state->ops.cop2_pgxp)` will be fine.

pcercuei commented on lightrec.h:
```diff
@@ -87,6 +87,9 @@ struct lightrec_mem_map {
 	void (*enable_ram)(struct lightrec_state *state, _Bool enable);
+
+	/* Optional, for PGXP support */
+	void (*cop2_pgxp)(struct lightrec_state *state, u32 op, u32 r);
```
What about `mtc2_notify` as the callback name? Because it's only for MTC2/CTC2, and there could be other potential uses outside PGXP.

In any case - you will need to rebase this commit on top of master, since I added a new callback in the lightrec_ops and it probably doesn't rebase cleanly.

ZachCook replied:
It's used for mfc too, so mfc_mtc_notify or maybe mxc_notify, or cop2_notify? Any preferences?

pcercuei replied:
Oh yeah you are right. Then I'd prefer `cop2_notify` and you can move the callback below or above `cop2_op` just to make sure that they are grouped.

ZachCook committed:
Add optional cop2_notify callback to lightrec_ops

The callback is given the full opcode and the data that was read or written to or from the cop2 register

The callback is expected to decode the opcode, for example to ignore LWC2/SWC2 and determine which of the MFC/CFC/MTC/CTC functions was called

This will be used to provide the opcode and return/data value to PGXP_GTE_[MFC/CFC/MTC/CTC] functions in beetle-psx that was lost when cop2 became internal to lightrec

Signed-off-by: Zachary Cook <zachcook1991@gmail.com>

pcercuei commented on lightrec.c:
```diff
@@ -432,22 +432,26 @@ u32 lightrec_mfc(struct lightrec_state *state, union code op)
+		case 27:
+		case 29:
+		case 30:
+			val = (u32)(s16)val;
```
missing
```
break;
default:
break;
```

or:

```
fallthrough;
default:
break;
```

pcercuei commented:
@ZachCook I can merge it now, but it looks quite incomplete to me. I think you overlooked the fact that the emitted code will directly manipulate Lightrec's CP2 registers when possible.

ZachCook commented:
> @ZachCook I can merge it now, but it looks quite incomplete to me. I think you overlooked the fact that the emitted code will directly manipulate Lightrec's CP2 registers when possible.

Ah, I didn't know that was being done, would it be better to not use cop2 emitted code at all when cop2_notify is provided or call cop2_notify from emitted code?

pcercuei commented:
If you look at `rec_mtc0` in emitter.c it already has a path where it calls C for when a block is in BIOS (to handle the RAM being enabled/disabled). I think what you can do is call `rec_mtc` in the `rec_cp2_basic_*` emitters if you detect that `state->ops.cop2_notify` has been provided.

ZachCook committed:
Revert "lightrec: Remove MFC wrapper", for cop2_notify

This reverts commit 01c3d9ee6dd8891a64fe956ebdd433f76dde3ef8.

The wrapper is needed to restore rec_mfc, so cop2_notify can be called
by the emitted code
Signed-off-by: Zachary Cook <zachcook1991@gmail.com>

ZachCook committed:
emitter: Call rec_mfc and rec_mtc from rec_cp2_basic_* for cop2_notify

rec_cp2_basic_* functions must fallback to the C wrapper functions for
mfc/mtc when the cop2_notify callback is provided

The mtc/mfc in the C wrappers supports the cop2_notify callback which
will be used by beetle-psx for PGXP support

Signed-off-by: Zachary Cook <zachcook1991@gmail.com>

pcercuei commented:
Can you improve a bit the commit messages of the last two commits? It doesn't need much, just explain why you're doing these changes. And keep it <= 72 characters wide.
